### PR TITLE
Replace copy constructor with newBuilder() API.

### PR DESCRIPTION
**Pull request by swankjesse**
This new method is polymorphic.

### CODE REVIEW

File: wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 1]
```diff
@@ -630,7 +631,7 @@ private MethodSpec messageAdapterDecode(NameAllocator nameAllocator, MessageType
     result.beginControlFlow("default:");
     result.addStatement("$T fieldEncoding = reader.peekFieldEncoding()", FieldEncoding.class);
     result.addStatement("$T value = fieldEncoding.rawProtoAdapter().decode(reader)", Object.class);
-    result.addStatement("builder.addUnknownField(tag, fieldEncoding, value);");
```
Comment (JakeWharton) [inline wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java]: Should JavaPoet detect and log a warning or throw when you do this?

**Comment by JakeWharton** [comment]
I think this make the right trade-offs. Plus one less thing using reflection!

File: wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java (hunk 1 above)
Comment (swankjesse) [inline wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java]: https://github.com/square/javapoet/issues/333

**Comment by swankjesse** [comment]
Yep. Migrating to this will be a lot of work!

**Pull request merged**